File: benchmarks/run_all.py
"""Run the full Python baseline suite and emit JSON + a sorted table.

Usage:
    uv run python -m benchmarks.run_all             # all groups
    uv run python -m benchmarks.run_all serialize   # subset by keyword
    uv run python -m benchmarks.run_all --json OUT.json

Writes ``benchmarks/results/python_baseline.json`` by default (machine/env
metadata + every Result) so the optimization pass can diff before/after.
"""

from __future__ import annotations

import json
import platform
import sys
import time
from pathlib import Path

from benchmarks._harness import Result, summarize

_RESULTS_DIR = Path(__file__).parent / "results"


def _env() -> dict:
    import capnweb

    info = {
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
        "python_version": sys.version.split()[0],
        "python_impl": platform.python_implementation(),
        "platform": platform.platform(),
        "processor": platform.processor(),
        "machine": platform.machine(),
        "capnweb_version": getattr(capnweb, "__version__", "?"),
    }
    try:
        import subprocess

        info["cpu_brand"] = subprocess.check_output(
            ["sysctl", "-n", "machdep.cpu.brand_string"], text=True
        ).strip()
        info["ncpu"] = subprocess.check_output(
            ["sysctl", "-n", "hw.ncpu"], text=True
        ).strip()
    except Exception:
        pass
    return info


def collect(filter_kw: str | None = None) -> list[Result]:
    from benchmarks import (
        bench_serialize,
        bench_rpc,
        bench_streams,
        bench_tables,
    )

    modules = {
        "serialize": bench_serialize,
        "rpc": bench_rpc,
        "streams": bench_streams,
        "tables": bench_tables,
    }
    results: list[Result] = []
    for key, mod in modules.items():
        if filter_kw and filter_kw not in key:
            continue
        print(f"[run_all] running {key} ...", file=sys.stderr, flush=True)
        results += mod.run()
    return results


def main(argv: list[str]) -> None:
    filter_kw = None
    out_path = _RESULTS_DIR / "python_baseline.json"
    args = list(argv)
    if "--json" in args:
        i = args.index("--json")
        out_path = Path(args[i + 1])
        del args[i : i + 2]
    if args:
        filter_kw = args[0]

    results = collect(filter_kw)

    # Sort so the table groups cleanly.
    results.sort(key=lambda r: (r.group, r.name))
    print(summarize(results))
    print()
    for r in results:
        if r.extra:
            print(f"  {r.name}: {r.extra}")

    _RESULTS_DIR.mkdir(exist_ok=True)
    payload = {"env": _env(), "results": [r.as_dict() for r in results]}
    out_path.write_text(json.dumps(payload, indent=2))
    print(f"\n[run_all] wrote {out_path}", file=sys.stderr)


if __name__ == "__main__":
    main(sys.argv[1:])

File: benchmarks/ts_compare/run_all.ts
// Run the TS reference suite and emit JSON + a table, mirroring
// benchmarks/run_all.py. Run from this directory (node_modules is symlinked to
// the interop install of capnweb 0.9.0 + tsx):
//
//   npx tsx run_all.ts               # all
//   npx tsx run_all.ts serialize     # subset by keyword
//   npx tsx run_all.ts --json out.json

import { writeFileSync, mkdirSync, readFileSync } from "node:fs";
import { execSync } from "node:child_process";
import * as os from "node:os";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __here = dirname(fileURLToPath(import.meta.url));
import { Result, printTable } from "./harness.js";
import { run as runSerialize } from "./bench_serialize.js";
import { run as runRpc } from "./bench_rpc.js";
import { run as runStreams } from "./bench_streams.js";

function env(): Record<string, unknown> {
  let cpu = "";
  try { cpu = execSync("sysctl -n machdep.cpu.brand_string").toString().trim(); } catch { /* non-macos */ }
  let capnwebVersion = "?";
  try {
    capnwebVersion = JSON.parse(
      readFileSync(join(__here, "node_modules/capnweb/package.json"), "utf8"),
    ).version;
  } catch { /* ignore */ }
  return {
    timestamp: new Date().toISOString(),
    node_version: process.version,
    v8_version: process.versions.v8,
    platform: `${os.type()} ${os.release()}`,
    arch: os.arch(),
    cpu_brand: cpu,
    ncpu: os.cpus().length,
    capnweb_version: capnwebVersion,
  };
}

async function main() {
  const args = process.argv.slice(2);
  let outPath = join(__here, "..", "results", "ts_baseline.json");
  const ji = args.indexOf("--json");
  if (ji >= 0) { outPath = args[ji + 1]; args.splice(ji, 2); }
  const filter = args[0];

  const groups: Record<string, () => Result[] | Promise<Result[]>> = {
    serialize: runSerialize,
    rpc: runRpc,
    streams: runStreams,
  };

  let results: Result[] = [];
  for (const [key, fn] of Object.entries(groups)) {
    if (filter && !key.includes(filter)) continue;
    process.stderr.write(`[run_all] running ${key} ...\n`);
    results = results.concat(await fn());
  }

  results.sort((a, b) => (a.group + a.name).localeCompare(b.group + b.name));
  printTable(results);
  console.log();
  for (const r of results) {
    if (Object.keys(r.extra).length) console.log(`  ${r.name}: ${JSON.stringify(r.extra)}`);
  }

  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, JSON.stringify({ env: env(), results }, null, 2));
  process.stderr.write(`\n[run_all] wrote ${outPath}\n`);
}

main().then(() => process.exit(0)).catch((e) => { console.error(e); process.exit(1); });
